### PR TITLE
Chloromorph code update to flags instead of hardcoded traits

### PR DESCRIPTION
**Pull request by RandDenner**
#### Summary
None

#### Purpose of change

Update code added by PR #55399 ("Makes Chlormorph an explicit upgrade over Roots/Rooter") to utilize flag checks instead of hardcoded traits

#### Describe the solution

A few lines of code in two or three files.

Will affect mutation.cpp and morale.cpp
And possibly character.cpp

Uncertain if character.cpp needs updating. Will look at it and see.

#### Testing

Seems to be working properly.

#### Additional context

Requested by @I-am-Erk

### CODE REVIEW

**Comment by RandDenner** [comment]
Hmm... I'm uncertain of something in morale.cpp

OP from #55399 added chloromorph to boolean condition check but may have missed something.
Lines 1074 to 1094:
```
void player_morale::update_constrained_penalty()
{
    const auto bp_pen = [ this ]( bodypart_id bp, int pen ) -> int {
        return ( body_parts[bp].covered > 0 ) ? pen : 0;
    };
    int pen = 0;

    if( has_mutation( trait_FLOWERS ) ) {
        pen += bp_pen( bodypart_id( "head" ), 10 );
    }
    if( has_mutation( trait_ROOTS1 ) || has_mutation( trait_ROOTS2 ) ||
        has_mutation( trait_ROOTS3 ) || has_mutation( trait_CHLOROMORPH ) ) {
        pen += bp_pen( bodypart_id( "foot_l" ), 5 );
        pen += bp_pen( bodypart_id( "foot_r" ), 5 );
    }
    if( has_mutation( trait_LEAVES2 ) || has_mutation( trait_LEAVES3 ) ) {
        pen += bp_pen( bodypart_id( "arm_l" ), 5 );
        pen += bp_pen( bodypart_id( "arm_r" ), 5 );
    }
    set_permanent( MORALE_PERM_CONSTRAINED, -std::min( pen, 10 ) );
}
```
OP added `has_mutation( trait_CHLOROMORPH ) )`
but no added entry in lines 253 to 287
```
{
    using namespace std::placeholders;
    // Cannot use 'this' because the object is copyable
    const auto set_optimist       = std::bind( &player_morale::set_permanent, _1, MORALE_PERM_OPTIMIST,
                                    _2, nullptr );
    const auto set_badtemper      = std::bind( &player_morale::set_permanent, _1, MORALE_PERM_BADTEMPER,
                                    _2, nullptr );
    const auto set_numb           = std::bind( &player_morale::set_permanent, _1, MORALE_PERM_NUMB,
                                    _2, nullptr );
    const auto set_stylish        = std::bind( &player_morale::set_stylish, _1, _2 );
    const auto update_constrained = std::bind( &player_morale::update_constrained_penalty, _1 );
    const auto update_masochist   = std::bind( &player_morale::update_masochist_bonus, _1 );

    mutations[trait_OPTIMISTIC]    = mutation_data(
                                         std::bind( set_optimist, _1, 9 ),
                                         std::bind( set_optimist, _1, 0 ) );
    mutations[trait_BADTEMPER]     = mutation_data(
                                         std::bind( set_badtemper, _1, -9 ),
                                         std::bind( set_badtemper, _1, 0 ) );
    mutations[trait_NUMB]          = mutation_data(
                                         std::bind( set_numb, _1, -1 ),
                                         std::bind( set_numb, _1, 0 ) );
    mutations[trait_STYLISH]       = mutation_data(
                                         std::bind( set_stylish, _1, true ),
                                         std::bind( set_stylish, _1, false ) );
    mutations[trait_FLOWERS]       = mutation_data( update_constrained );
    mutations[trait_ROOTS1]        = mutation_data( update_constrained );
    mutations[trait_ROOTS2]        = mutation_data( update_constrained );
    mutations[trait_ROOTS3]        = mutation_data( update_constrained );
    mutations[trait_LEAVES2]       = mutation_data( update_constrained );
    mutations[trait_LEAVES3]       = mutation_data( update_constrained );
    mutations[trait_MASOCHIST]     = mutation_data( update_masochist );
    mutations[trait_MASOCHIST_MED] = mutation_data( update_masochist );
    mutations[trait_CENOBITE]      = mutation_data( update_masochist );
}
```
between trait_ROOTS3 and trait_LEAVES2.

Oversight?

Not certain enough of the code to be sure if I should add it.

Also uncertain but strongly thinking that I should not update the trait code to flags in this file.
It's not used in any of this code as yet. May be a bad idea atm, especially given my n00b knowledce of C++ coding...

**Comment by RandDenner** [comment]
In character.cpp, there are 8 instances of trait_CHLOROMORPH

In the defines (390): `static const trait_id trait_CHLOROMORPH( "CHLOROMORPH" );`
(which I may have to leave, adding a line for `static const json_character_flag json_flag_CHLOROMORPH( "CHLOROMORPH" );`  because of the possible issues if I mess with the 2 has_active_mutation( trait_CHLOROMORPH ) lines. See end of this comment for thinking on this.)

4606 to 4614:
```
    if( has_activity( ACT_TREE_COMMUNION ) ) {
        // Much of the body's needs are taken care of by the trees.
        // Hair Roots don't provide any bodily needs.
        if( has_trait( trait_ROOTS2 ) || has_trait( trait_ROOTS3 ) || has_trait( trait_CHLOROMORPH ) ) {
            rates.hunger *= 0.5f;
            rates.thirst *= 0.5f;
            rates.fatigue *= 0.5f;
        }
    }
```
4956: `bool plantsleep = has_trait( trait_CHLOROMORPH );`

7641 to 7649:
```
void Character::rooted_message() const
{
    bool wearing_shoes = footwear_factor() == 1.0;
    if( ( has_trait( trait_ROOTS2 ) || has_trait( trait_ROOTS3 ) || has_trait( trait_CHLOROMORPH ) ) &&
        get_map().has_flag( ter_furn_flag::TFLAG_PLOWABLE, pos() ) &&
        !wearing_shoes ) {
        add_msg( m_info, _( "You sink your roots into the soil." ) );
    }
}
```
7656 to 7673:
```
{
    double shoe_factor = footwear_factor();
    if( ( has_trait( trait_ROOTS2 ) || has_trait( trait_ROOTS3 ) || has_trait( trait_CHLOROMORPH ) ) &&
        get_map().has_flag( ter_furn_flag::TFLAG_PLOWABLE, pos() ) && shoe_factor != 1.0 ) {
        int time_to_full = 43200; // 12 hours
        if( has_trait( trait_ROOTS3 ) || has_trait( trait_CHLOROMORPH ) ) {
            time_to_full += -14400;    // -4 hours
        }
        if( x_in_y( 96, time_to_full ) ) {
            vitamin_mod( vitamin_iron, 1 );
            vitamin_mod( vitamin_calcium, 1 );
            mod_healthy_mod( 5, 50 );
        }
        if( get_thirst() > -40 && x_in_y( 288, time_to_full ) ) {
            mod_thirst( -1 );
        }
    }
}
```
and 8918 to 8921
```
        if( ( has_trait( trait_ROOTS3 ) || has_trait( trait_CHLOROMORPH ) ) &&
            here.has_flag( ter_furn_flag::TFLAG_DIGGABLE, pos() ) ) {
            movecost += 10 * footwear_factor();
        }
```
The oddballs are:
7994: `} else if( has_active_mutation( trait_CHLOROMORPH ) ) {`
and
9509: `bool activechloro = has_active_mutation( trait_CHLOROMORPH );`

Which I'm thinking maybe don't get converted to flag check versions?
But I don't know if `has_active_mutation` can deal with `json_flag_CHLOROMORPH`.

Gonna need some experienced coding advice on this one.

**Comment by RandDenner** [comment]
Added `    mutations[trait_CHLOROMORPH]   = mutation_data( update_constrained );`
to the function `void player_morale::mutation_data::clear()` (line 241)
that the OP missed.

**Comment by RandDenner** [comment]
I think this is all correct.
Seems to be working properly afaict.

File: src/mutation.cpp [hunk 1]
```diff
@@ -765,7 +765,7 @@ void Character::activate_mutation( const trait_id &mut )
             return;
         }
 
-        if( has_trait( trait_ROOTS2 ) || has_trait( trait_ROOTS3 ) || has_trait( trait_CHLOROMORPH ) ) {
+        if( has_flag( json_flag_ROOTS2 ) || has_flag( json_flag_ROOTS3 ) || has_flag( json_flag_CHLOROMORPH ) ) {
```
Comment (kevingranade) [inline src/mutation.cpp:768]:
```suggestion
        if( has_flag( json_flag_ROOTS2 ) || has_flag( json_flag_ROOTS3 ) ||
            has_flag( json_flag_CHLOROMORPH ) ) {
```

**Pull request merged**